packages/vscode-extension/src/webview/views/EnvEditor.tsx

```diff
@@ -0,0 +1,163 @@
+        </VscodeTableHeader>
+        <VscodeTableBody slot="body">
+          {entries.map(([key, value]) => (
+            <VscodeTableRow key={key} onDoubleClick={() => startEditing(key)}>
+              <VscodeTableCell>
+                {editingKey === key || key === null ? (
+                  <VscodeTextfield
+                    placeholder="Key"
+                    value={editingKeyValue}
+                    data-no-submit
+                    onInput={(e) => setEditingKeyValue((e.target as HTMLInputElement).value)}
+                    onKeyDown={(e) => {
+                      if (e.key === "Enter") {
+                        e.preventDefault();
+                        save();
+                      }
+                    }}
+                  />
+                ) : (
+                  key
+                )}
+              </VscodeTableCell>
+              <VscodeTableCell>
+                {editingKey === key || key === null ? (
+                  <VscodeTextfield
+                    placeholder="Value"
+                    value={editingValue}
+                    data-no-submit
+                    onInput={(e) => setEditingValue((e.target as HTMLInputElement).value)}
+                    onKeyDown={(e) => {
+                      if (e.key === "Enter") {
+                        e.preventDefault();
+                        save();
+                      }
+                    }}
+                  />
+                ) : (
+                  value
+                )}
+              </VscodeTableCell>
+              <VscodeTableCell>
+                <div className="env-editor-table-cell-actions">
+                  {editingKey === key || key === null ? (
+                    <>
+                      <Button disabled={!editingKeyValue} onClick={save}>
+                        OK
+                      </Button>
+                      <Button onClick={cancel} secondary>
+                        Cancel
+                      </Button>
+                    </>
+                  ) : (
+                    <>
+                      <ToolbarButton
+                        onClick={() => startEditing(key)}
+                        title="Edit Environment Variable">
+                        <span className="codicon codicon-edit" />
+                      </ToolbarButton>
+                      <ToolbarButton
+                        onClick={() => deleteEntry(key)}
+                        title="Remove Environment Variable">
+                        <span className="codicon codicon-close" />
+                      </ToolbarButton>
+                    </>
+                  )}
+                </div>
+              </VscodeTableCell>
+            </VscodeTableRow>
+          ))}
+        </VscodeTableBody>
+      </VscodeTable>
+      {!showAddForm && <Button onClick={() => startEditing(null)}>Add Variable</Button>}
+      <input type="hidden" name="env" value={JSON.stringify(env)} />
```

Comment:
I don't know how I feel about this vs having a controlled prop / returning the value via callback.

Reply:
this is mainly for consistency reasons with all the other fields that we now collect via the FormData. Otherwise this would be the only one controlled component which would feel off.